### PR TITLE
implement basic check for auto burn in prompt

### PR DESCRIPTION
**Pull request by mark-fathom**
Detect a not-awful place to set the burn-in threshold.

### CODE REVIEW

File: src/ts/ui/runner/burninprompt.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      const secondHalf = series.slice(midpoint);
+      /* get the stddev of the second half */
+      const sum = secondHalf.reduce((tot, c)=>tot+(c||0), 0);
+      const avg = sum / secondHalf.length;
+      const sumDeltaSq = secondHalf.reduce((tot, c)=>{
+        const d = c - avg;
+        return tot + d * d;
+      }, 0);
+      const stdDev = Math.sqrt(sumDeltaSq/secondHalf.length);
+      for (let i = midpoint - 1; i >= 0; i--) {
```
Comment (patvarilly) [inline src/ts/ui/runner/burninprompt.ts:21]: I'd even start at `series.length - 1` instead of `midpoint - 1`; there's no guarantee that the second half of the series is already past some rough burn-in cutoff

File: src/ts/ui/runner/runui.ts [hunk 2]
```diff
@@ -434,14 +445,32 @@ export class RunUI extends UIScreen {
...
+        this.muStarCanvas.setData(muudStar, kneeIndex, mccIndex, hideBurnIn, sampleIndex);
+        serieses.push(muudStar);
       }
-      this.TCanvas.setData(totalBranchLengthHist.map(t=>t/DAYS_PER_YEAR), kneeIndex, mccIndex, hideBurnIn, sampleIndex);
+      this.TCanvas.setData(totalLengthYear, kneeIndex, mccIndex, hideBurnIn, sampleIndex);
       this.mutCountCanvas.setData(numMutationsHist, kneeIndex, mccIndex, hideBurnIn, sampleIndex);
-      this.popGrowthCanvas.setData(popGHist.map(g=>POP_GROWTH_FACTOR/g), kneeIndex, mccIndex, hideBurnIn, sampleIndex);
+      this.popGrowthCanvas.setData(popHistGrowth, kneeIndex, mccIndex, hideBurnIn, sampleIndex);
+
+      if (this.pythia && this.pythia.kneeIndex <= 0) {
```
Comment (patvarilly) [inline src/ts/ui/runner/runui.ts:465]: It would be even better to keep adjusting the proposed knee index as long as the user hasn't specified one explicitly.  Here's a screenshot of how the current setup doesn't make a great guess to begin with, but later has enough information to make a better guess (I ran the demo and set "Steps per sample" to 100,000; the cursor is around where I would have placed the burn-in cutoff manually)
Screenshot from 2024-10-22 09-46-43

**Review by patvarilly — CHANGES_REQUESTED** [review]
Looking good !  My main ask would be to keep updating the suggested knee index as the run continues (unless the user has clicked something specifically).  Would it be possible to somehow visually distinguish a "suggested knee" from a "user-selected knee" (e.g., a dashed line, or a small note somewhere saying "Suggested burn-in cutoff") ?

File: src/ts/ui/runner/burninprompt.ts (hunk 1 above)
Comment (mark-fathom) [reply src/ts/ui/runner/burninprompt.ts]: Just pushed up a change to start at `last index - 1`, so that we have at least one sample after the knee, but if that's too arbitrary, let me know.

File: src/ts/ui/runner/runui.ts (hunk 2 above)
Comment (mark-fathom) [reply src/ts/ui/runner/runui.ts:465]: Trickier, but I'll take a peek.

Comment (mark-fathom) [reply src/ts/ui/runner/runui.ts:465]: hrm, if I never override the auto suggested knee, should we export the kneeIndex? Because if I reload it, currently there would be no way to distinguish between the auto-selected knee and a curated knee. We can do a bunch of juggling to unset the knee on export and then reset it, and auto set the knee on load, or we can add a flag to the export for something like `kneeIsManuallySet` or `kneeIsCurated` ?

**Comment by mark-fathom** [comment]
Question: if the user never overrides the auto suggested knee, should we export the kneeIndex? Because on reload, there would be no way to distinguish between the auto-selected knee and a curated knee. We can do a bunch of juggling to unset the knee on export and then reset it, and auto set the knee on load, or we can add a flag to the export for something like `kneeIsManuallySet` or `kneeIsCurated` ?

**Comment by patvarilly** [comment]
Good point!  I would go for not saving the knee on export, then auto-set it
on load if it hasn't yet been chosen.  That leaves your hands free to
implement other suggestions for the knee in the future, and also adds a
good suggested knee to previously saved runs where a user "accidentally"
forgot to set a knee.

File: src/ts/ui/runner/runui.ts (hunk 2 above)
Comment (mark-fathom) [reply src/ts/ui/runner/runui.ts:465]: Hrm, consistently updating the knee from the latest sample results in only a few sampled trees. Here's a long run of 1777 samples, and you can see the knee allows only 4: 
Screen Shot 2024-10-22 at 9 26 44 AM

Exporting and reloading the file with the the acceptable range at 3 std deviations rather than just 2, we get 53 samples, at 4 std dev we get 56, 5 -> 1312. 

Hrm, if we repeatedly calculate the knee, then I retract my comment about needing to set the limit at 2 std dev. That's too restrictive, at least for this data set. 5 doesn't look that bad while in progress: 
Screen Shot 2024-10-22 at 9 53 57 AM

@patvarilly what do you think? 

N.B. still have to work out the mechanics of distinguishing between the auto and user selected knee.

**Comment by patvarilly** [comment]
So, there's a good answer, and there's a fudge.

The good answer is as follows.  If you were sure that the samples in the
traces were independent, then you expect some samples to be more than
1,2,3,... std devs away from the mean just by chance (the usual numbers are
that 68% of samples are within 1 s.d., 95% are within 2 s.d., 99.7% are
within 3 s.d, 99.994% are within 4 s.d. and 99.99994% are within 5 s.d.).
If a fraction x is outside the test range, then the expected number of
in-range samples you can expect to see before you first hit an out-of-range
sample follows a Geometric distribution (
https://en.wikipedia.org/wiki/Geometric_distribution) with p = x, which has
mean 1-p / p.  So for 1,2,3,4,5 s.d.s, that's respectively 2.1, 19, 332,
16.7k and 16.7M. You could go a little further and pick an s.d. threshold
such that there's, say, a 99% chance that you won't cut off any of the N
samples in your run just by chance.  But that's probably overkill for
Delphy web.  Instead, I'd go with "3 or 4 std devs" and call it a day.

The fudge is that you *don't* have independent samples in the traces yet.
You'd need to calculate the Effective Sample Size (ESS) and then calculate
the std. dev. of the traces using that instead of the actual number of
samples in the trace (i.e., std dev = sqrt([sum_i (x_i - mean_x)^2] / ESS)
instead of sqrt([...] / N)  (there's also a `-1` correction in the
denominator that, between friends, we'll ignore).  If you *did* have ESSs
handy (and I hope we do some day), then that together with 3-4 std devs
would seem like a great heuristic.  Since we don't yet, going up to 5 "std
devs" seems a good compromise for now.

Tl;dr: a cutoff of 5 s.d.'s seems reasonable to me for now

**Review by patvarilly — APPROVED** [review]
This works incredibly well !!  Nice !  The suggested knee is usually very reasonable, and will certainly spare most users from making an egregious error in using Delphy.  I made two little changes after playing with this: I excluded two observables (number of mutations & doubling time) from the estimation, because their equilibrium distribution is very non-Gaussian, and also added a bit that advances the suggestion forward from the last point that's 5 s.d.s away to the first point after that that's within 2 s.d.s of the mean, to avoid grabbing the "tail end" of the burn-in period.  I ran this against many of our benchmarks (Ebola, Zika, SC2, the simulations) and various sampling rates, and the suggested knee is now almost always where I would put it manually.

**Comment by mark-fathom** [comment]
Awesome, thank you for updating the algorithm!
